File: archive/g/go/transpose-matrix.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	if len(os.Args) < 4 {
		fmt.Println("Usage: please enter the dimension of the matrix and the serialized matrix")
		return
	}

	cols, err1 := strconv.Atoi(os.Args[1])
	rows, err2 := strconv.Atoi(os.Args[2])
	matrixStr := os.Args[3]

	if err1 != nil || err2 != nil || cols <= 0 || rows <= 0 || matrixStr == "" {
		fmt.Println("Usage: please enter the dimension of the matrix and the serialized matrix")
		return
	}

	matrixStr = strings.ReplaceAll(matrixStr, " ", "")
	matrixElements := strings.Split(matrixStr, ",")
	if len(matrixElements) != cols*rows {
		fmt.Println("Usage: please enter the dimension of the matrix and the serialized matrix")
		return
	}

	matrix := make([][]int, rows)
	for i := range matrix {
		matrix[i] = make([]int, cols)
	}

	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			num, err := strconv.Atoi(matrixElements[i*cols+j])
			if err != nil {
				fmt.Println("Usage: please enter the dimension of the matrix and the serialized matrix")
				return
			}
			matrix[i][j] = num
		}
	}

	transposed := transpose(matrix)
	printMatrix(transposed)
}

func transpose(matrix [][]int) [][]int {
	if len(matrix) == 0 {
		return [][]int{}
	}

	transposed := make([][]int, len(matrix[0]))
	for i := range transposed {
		transposed[i] = make([]int, len(matrix))
	}

	for i := range matrix {
		for j := range matrix[i] {
			transposed[j][i] = matrix[i][j]
		}
	}

	return transposed
}

func printMatrix(matrix [][]int) {
	var result []string
	for _, row := range matrix {
		for _, val := range row {
			result = append(result, strconv.Itoa(val))
		}
	}
	fmt.Println(strings.Join(result, ", "))
}

File: archive/f/f-sharp/TransposeMatrix.fs
open System

module Matrix =
    let transpose (data: int[,]) =
        let rows = data.GetLength 0
        let cols = data.GetLength 1

        Array2D.init cols rows (fun x y ->
            data.[y, x]
        )

module Helpers =
    let parseInt (s: string) =
        match s.Trim() |> Int32.TryParse with
        | true, v -> Some v
        | _ -> None

    let parseMatrix (s: string) =
        s.Split(',', StringSplitOptions.RemoveEmptyEntries)
        |> Array.choose parseInt
        |> function
            | arr when arr.Length > 0 -> Some arr
            | _ -> None

    let parseArgs (argv: string[]) =
        match argv with
        | [| cStr; rStr; matrixStr |] ->
            match parseInt cStr, parseInt rStr, parseMatrix matrixStr with
            | Some c, Some r, Some values when values.Length = c * r ->
                Ok (c, r, values)

            | _ ->
                Error ()

        | _ ->
            Error ()

    let buildMatrix (c, r, values: int[]) =
        Array2D.init r c (fun i j ->
            values.[i * c + j]
        )

    let handleResult =
        function
        | Ok m ->
            m |> Seq.cast<int> |> Seq.map string |> String.concat ", " |> printfn "%s"
            0

        | Error () ->
            printfn "Usage: please enter the dimension of the matrix and the serialized matrix"
            1


[<EntryPoint>]
let main argv =
    argv
    |> Helpers.parseArgs
    |> Result.map Helpers.buildMatrix
    |> Result.map Matrix.transpose
    |> Helpers.handleResult